temporal-sdk/src/main/java/io/temporal/internal/external/ManualActivityCompletionClientImpl.java

```diff
@@ -202,45 +192,45 @@ public void fail(Throwable exception) {
                     .blockingStub()
                     .withOption(METRICS_TAGS_CALL_OPTIONS_KEY, metricsScope)
                     .respondActivityTaskFailedById(request));
-      } catch (StatusRuntimeException e) {
-        if (e.getStatus().getCode() == Status.Code.NOT_FOUND) {
-          throw new ActivityNotExistsException(activityId, e);
-        }
-        throw new ActivityCompletionFailureException(activityId, e);
       } catch (Exception e) {
-        throw new ActivityCompletionFailureException(activityId, e);
+        processException(e);
       }
     }
   }
 
+  @Override
+  public void sendHeartbeatRequest(Object details) {
+    RecordActivityTaskHeartbeatRequest.Builder request =
+        RecordActivityTaskHeartbeatRequest.newBuilder()
+            .setTaskToken(ByteString.copyFrom(taskToken))
+            .setNamespace(namespace)
+            .setIdentity(identity);
+    Optional<Payloads> payloads = dataConverter.toPayloads(details);
+    payloads.ifPresent(request::setDetails);
+    RecordActivityTaskHeartbeatResponse status;
+    status =
+        service
+            .blockingStub()
+            .withOption(METRICS_TAGS_CALL_OPTIONS_KEY, metricsScope)
+            .recordActivityTaskHeartbeat(request.build());
+    if (status.getCancelRequested()) {
```

Comment:
From what I see, it's better to move this part out from this method to the callers. Because both users catch this exception and process them anyway. They can process the status directly.